plugins/jobtype/src/azkaban/jobtype/HadoopTuningSecurePigWrapper.java

```diff
@@ -0,0 +1,181 @@
+import java.io.FileReader;
+import java.security.PrivilegedExceptionAction;
+import java.util.Properties;
+
+import org.apache.hadoop.security.UserGroupInformation;
+import org.apache.log4j.Logger;
+import org.apache.pig.tools.pigstats.PigStats;
+
+import azkaban.jobtype.pig.PigUtil;
+import azkaban.jobtype.tuning.TuningCommonConstants;
+import azkaban.jobtype.tuning.TuningErrorHandler;
+import azkaban.jobtype.tuning.TuningParameterUtils;
+import azkaban.utils.Props;
+
+
+/**
+ * This class represent wrapper for running pig job with tuning enabled.
+ */
+public class HadoopTuningSecurePigWrapper {
+
+  public static final String WORKING_DIR = "working.dir";
+
+  private static File pigLogFile;
+
+  private static Props props;
+
+  private static final Logger logger;
+
+  /**
+   * In case if job is failed because of auto tuning parameters, it will be retried with the best parameters
+   * we have seen so far. Maximum number of try is 2.
+   */
+  private static int maxJobRetry = 2;
+  private static int jobTryCount = 1;
+  /**
+   * Is job failed because of tuning parameters
+   */
+  private static boolean isTuningError = false;
+
+  private static final String TUNING_JOB_RETRY_COUNT = "tuning.job.retry.count";
+
+  static {
+    logger = Logger.getRootLogger();
+  }
+
+  /**
+   * This function runs the pig job with tuning enabled. In case tuning job fails in first try, it asks for
+   * last best parameter from tuning and re-run the job. In case tuning end point is not responding this job
+   * will run with default parameters.
+   * @param args
+   * @throws Throwable
+   */
+  public static void main(final String[] args) throws Exception {
+    Properties jobProps = HadoopSecureWrapperUtils.loadAzkabanProps();
+    Props initialJobprops = new Props(null, jobProps);
+    boolean retry = false;
+    boolean firstTry = true;
+    if (props.containsKey(TUNING_JOB_RETRY_COUNT)) {
+      maxJobRetry = props.getInt(TUNING_JOB_RETRY_COUNT);
+    }
+
+    while (jobTryCount <= maxJobRetry && (retry || firstTry)) {
+      jobTryCount++;
+      firstTry = false;
+      props = Props.clone(initialJobprops);
+      props.put(TuningCommonConstants.AUTO_TUNING_RETRY, retry + "");
+
+      TuningParameterUtils.updateAutoTuningParameters(props);
+
+      HadoopTuningConfigurationInjector.prepareResourcesToInject(props,
+          HadoopTuningSecurePigWrapper.getWorkingDirectory(props));
+
+      HadoopTuningConfigurationInjector.injectResources(props);
+
+      // special feature of secure pig wrapper: we will append the pig error file
+      // onto system out
+      pigLogFile = new File(System.getenv("PIG_LOG_FILE"));
+
+      try {
+        if (HadoopSecureWrapperUtils.shouldProxy(jobProps)) {
+          String tokenFile = System.getenv(HADOOP_TOKEN_FILE_LOCATION);
+          UserGroupInformation proxyUser = HadoopSecureWrapperUtils.setupProxyUser(jobProps, tokenFile, logger);
+          proxyUser.doAs(new PrivilegedExceptionAction<Void>() {
```

Comment:
Is the language level on Azkaban high enough (Java 8) to use a lambda here?

Reply:
Yes.

Reply:
In that case, this could be greatly simplified like `proxyUser.doAs(() -> runPigJob(args));`

Reply:
For now I will keep it as it is.